oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -73,39 +73,52 @@ pub struct Runtime {
 }
 
 impl Runtime {
-    /// Configure and run the protobuf specified Application [`Configuration`]. After creating a
-    /// [`Runtime`], calling [`Runtime::stop`] will send termination signals to nodes and wait for
-    /// them to terminate. This returns a [`RuntimeRef`] reference to the created runtime, and a
-    /// writeable [`Handle`] to send messages into the runtime. To receive messages, creating a new
-    /// channel and passing the write [`Handle`] into the runtime will enable messages to be
-    /// read back out.
-    pub fn configure_and_run(config: Configuration) -> Result<(RuntimeRef, Handle), OakStatus> {
-        let runtime = Runtime {
-            configurations: config.nodes,
+    /// Creates a [`Runtime`] instance but does not start executing any node.
+    pub fn create(configuration: Configuration) -> Self {
+        Self {
+            configuration,
             terminating: AtomicBool::new(false),
             channels: channel::ChannelMapping::new(),
             nodes: Mutex::new(HashMap::new()),
             next_node_reference: AtomicU64::new(0),
-        };
+        }
+    }
 
-        let runtime = RuntimeRef(Arc::new(runtime));
+    /// Converts a [`Runtime`] instance into a [`RuntimeRef`].
+    pub fn into_ref(self) -> RuntimeRef {
+        RuntimeRef(Arc::new(self))
+    }
+
+    /// Configures and runs the protobuf specified Application [`Configuration`].
+    ///
+    /// After starting a [`Runtime`], calling [`Runtime::stop`] will send termination signals to
+    /// nodes and wait for them to terminate.
+    ///
+    /// Returns a [`RuntimeRef`] reference to the created runtime, and a writeable [`Handle`] to
+    /// send messages into the runtime. To receive messages, creating a new channel and passing
+    /// the write [`Handle`] into the runtime will enable messages to be read back out.
+    pub fn run(self) -> Result<(RuntimeRef, Handle), OakStatus> {
+        let module_name = self.configuration.entry_module.clone();
+        let entrypoint = self.configuration.entrypoint.clone();
+
+        let runtime_ref = self.into_ref();
 
         // When first starting, we assign the least privileged label to the channel connecting the
         // outside world to the entry point node.
         let (chan_writer, chan_reader) =
-            runtime.new_channel(&oak_abi::label::Label::public_trusted());
+            runtime_ref.new_channel(&oak_abi::label::Label::public_trusted());
 
-        runtime.node_create(
-            &config.entry_module,
-            &config.entrypoint,
+        runtime_ref.node_create(
+            &module_name,
+            &entrypoint,
             // When first starting, we assign the least privileged label to the entry point node.
             &oak_abi::label::Label::public_trusted(),
             chan_reader,
         )?;
 
-        runtime.channel_close(chan_reader)?;
+        runtime_ref.channel_close(chan_reader)?;
```

Comment:
Currently it's impossible for this call to fail as the `chan_reader` is guaranteed to be correct. If code changes such that this can fail then the runtime would spawn but there would be no handle to talk to it, so yeah leaked/lost.

Reply:
If this fails, the only reference to the `Arc` will go out of scope immediately and the `Runtime` object will be deallocated, so yes, it will be consumed, but not leaked in the sense of "memory leak".

Reply:
Actually if the `node_create` above succeeded, the node will have a reference to the runtime and it will not be deallocated, so it will leak.

Perhaps the close could be written more explicitly as 

~~~rust
// Ignore the result as this cannot fail as we created the handle above, and it cannot have been closed previously.
let _ = channel_close(chan_reader);
~~~

Reply:
If `channel_close` can't fail, maybe `expect` that to be the case rather than `?`.

(And if it did fail, doesn't the new Node typically hold a second `RuntimeRef` copy, and the `Runtime` underneath that holds a ref back to the Node instance, meaning things are leaked?)

Reply:
Replaced with `expect` + a comment now.

Not sure about the memory leak issue, is there anything else we should do here, or is this enough for now?

It does scare me a bit that it's easy enough to accidentally have cycles and end up leaking, not sure what to do about it, perhaps we should be using weak references in some cases?

Reply:
It can't fail so it can't actually leak here its just hypothetical, but we could probably make all the references passed to nodes weak pointers (and possibly call stop on drop of Runtime).

Reply:
Yeah, I think using weak refs would make sense, since we don't really care about counting incoming references from nodes to the runtime, right?

Reply:
Yeah sounds reasonable to me